File: BasicCal2.java
/*

Passed All Test Cases 

Time-> O(2N) -> O(N), N is # of characters in string
Space-> O(N), N is # of characters in string 

*/
class Solution {
    public int calculate(String s) {
        //cleaning string 
        s = s.replace(" ", "") ; 
        
        int value = 0 ; 
        char sign = '+' ; 
        
        int ans = 0 ; 
        
        Stack<Integer> stack = new Stack<>() ; 
        
        for (int i = 0 ; i < s.length() ; i++) {
            
            char check = s.charAt(i) ; 
            
            if(Character.isDigit(check)) {
                value = value * 10 + (check - '0') ; 
            }
            
            if(!Character.isDigit(check) || (i == s.length() - 1)) {
                
                if(sign == '+') {
                    stack.push(value) ; 
                } 
                else if(sign == '-') {
                    stack.push(-value) ; 
                } 
                else if (sign == '*') {
                    stack.push(stack.pop() * value) ; 
                } 
                else if (sign == '/') {
                    stack.push(stack.pop() / value) ; 
                }
                
                sign = check ; 
                
                value = 0 ; 
            }
        }
            
            while (!stack.isEmpty()) {
                ans += stack.pop() ; 
            }
            
            return ans ; 
        }
}

File: basicCal2.py
#Time complexity: O(n)
#Space complexity: O(1)
class Solution:
    def calculate(self, s: str) -> int:
        
        st = []
        lastSign = '+'
        currNum, res, n = 0, 0, len(s)
        
        
        for i in range(len(s)):
            char = s[i]
            if char.isdigit():
                currNum = currNum * 10 + (ord(char)-ord('0'))
                
            if ((not char.isdigit()) and char != ' ') or i == n-1:
                if lastSign == '+':
                    st.append(currNum)
                elif lastSign == '-':
                    st.append(-currNum)
                elif lastSign == '*':
                    st.append(st.pop() * currNum)
                elif lastSign == '/':
                    st.append(int(st.pop() / currNum))
                    
                lastSign = char
                currNum = 0
        
        while st:
            res += st.pop()
        
        return res
            
                    
        
                
        
        
